### DISCUSSION TITLE
pitch: Extending Application Statuses in RoundImplementation

### DISCUSSION THREAD

**Discussion by 0xKurt** [allo-protocol/aips#26] [New Ideas]
Inspired by a previous discussion (#11), I propose extending the application statuses in our round smart contract to provide greater granularity and flexibility.

Currently, our round smart contract utilizes a 2-bit system to store application statuses. However, as our platform evolves and new requirements emerge, it has become apparent that the existing status implementation lacks the necessary granularity and flexibility. Therefore, I propose extending the application statuses to a 4-bit system to accommodate a wider range of states and provide clearer information about the application's lifecycle.

In the existing implementation, the application statuses are defined as follows:
```
0: Pending
1: Approved
2: Rejected
3: Canceled
```
However, the usage of 0 to represent a pending application can be ambiguous and unreliable when relying on chain data. Additionally, it does not clearly distinguish between an absence of an application and a pending status.

I propose introducing an extended set of application statuses to address the limitations of the current implementation. Here's a suggested set of application statuses using a 4-bit system:
```
0: No application
1: Pending
2: Approved
3: Rejected
4: Approved, but Reapplied
5: Rejected, but Reapplied
6: [Additional Status]
n: ...
```
By introducing new statuses, we can better track the lifecycle of applications and provide more meaningful information to participants. The inclusion of "Approved, but Reapplied" and "Rejected, but Reapplied" allows us to differentiate between initial approvals/rejections and subsequent reapplications.

**Comment by lmcorbalan** [comment]
@KurtMerbeth What do you think about changing the bitmap approach?
e.g. Having a `status` property on the `Application` struct could make the code more readable, easy to understand and provide even greater granularity and flexibility.

**Reply by 0xKurt** [reply]
If you run a round with hundreds of applications and you have to iterate over multiple applications to update their status, it is very gas consuming.

Using the bitmap just saves a bunch of gas. I agree on what you said, but it would be a really expensive approach. (Especially for mainnet rounds)

**Reply by lmcorbalan** [reply]
Oh I see! That explains it all, agree

**Comment by nfrgosselin** [comment]
@KurtMerbeth I'm supportive of expanding the scope of application statuses that the protocol can handle. I'm curious — would the scope of this AIP also include creating a standard set of application statuses? I'm torn between allowing the flexibility of customizable statuses with also wanting to preserve some level of data usability.

**Reply by 0xKurt** [reply]
We def should have a standard set, else it will be messi to build frontends on top of it.

**Reply by codenamejason** [reply]
I agree, we should not allow for 'custom' statuses and use a standard set. A later AIP could address changes to any of the proposed status's.

**Comment by boudra** [comment]
@KurtMerbeth How would the reapplied statuses be used in practice? at the moment project application history is available to be queried, if a project applies to a round and they have application index 1, that application gets rejected, then they reapply with application index 2 that is accepted, this would be another application right? so you would end up basically with two applications:

- app 1: rejected
- app 2: accepted

With this information you know that the project was accepted in their second application, and this would be the source of truth, or are you thinking that we would update an application instead of creating a new one?

**Reply by 0xKurt** [reply]
In an ideal world, every project has just one status.
That's necessary to make it easier to build on top of the protocol. (And it's already wanted)

At this point it makes sense to add these application statuses, since you can only have one

**Reply by lmcorbalan** [reply]
There may be some escenarios where a project apply multiple times to the same program. That would be posible in the Direct Grant Rounds.
e.g. Let say I as a builder have multiple Ideas for contributing with Gitcoin that created a Direct Grant round, so I should be able to apply multiple times and each application should follow its own path

**Reply by 0xKurt** [reply]
But wouldn't be every idea a new project?
Users can have multiple projects, every project has a description metapointer, which is send with the application.

If the user wants to apply with different ideas, he would need to edit his project description every time, or just create a new project/idea - which makes most sense for me

**Reply by thelostone-mc** [reply]
> e.g. Let say I as a builder have multiple Ideas for contributing with Gitcoin that created a Direct Grant round, so I should be able to apply multiple times and each application should follow its own path

It does feel like a corner cause and it might be fair to state that if you are applying for a different idea -> ideally it should be it's project which would have it's own owners (not necessarily the parent projects owners). 

When a project applies -> a snapshot of the application is taken and sent to the round. 
If you are applying for a different idea -> ideally the application should contain all information around the idea and not the parent.

> Multiple Ideas for contributing with Gitcoin that created a Direct Grant round

Same usecase but if you wanted to apply to 2 rounds, it would be easier to have it as a separate project with it's own metadata as opposed filling out the custom application twice

**Reply by thelostone-mc** [reply]
>  How would the reapplied statuses be used in practice

Yup for that to happen, we would have to assume 1 project can submit 1 application to the round and the status can be toggled.
This would simplify the application process and allow easier tracking of history of the project.
Again this would only work if we allow applyToRound via the registry

**Reply by meglister** [reply]
I think that @lmcorbalan 's case is actually the "normal" case rather than an edge case for direct grants. It's very common for a direct grants program to have an open RFP (submit a great idea you want to build) and for projects to submit multiple applications/ideas for it.

**Reply by lmcorbalan** [reply]
> It does feel like a corner cause and it might be fair to state that if you are applying for a different idea -> ideally it should be it's project which would have it's own owners (not necessarily the parent projects owners).
> 
> Same usecase but if you wanted to apply to 2 rounds, it would be easier to have it as a separate project with it's own metadata as opposed filling out the custom application twice

For me as a builder (in a direct round context), I don't think it is a good idea to create a different project for each great idea I have because it wont allow me to grow a reputation. 

The way I see it, the meaning of a `Project` depends on the context, for QF Round it represents the combination of `Builder+Idea` and on the context of Direct Round it is just `Builder` and the `Application` is the `Idea`. Who knows what it would mean on a third context.

**Reply by 0xKurt** [reply]
I understand the need for a project to apply multiple times to a round, and I also acknowledge the concern about reputation, but having a project apply with different ideas doesn't feel quite right to me.
 
 In my opinion, a possible and clean solution could be to introduce the concept of an organization. This could be represented as follows:
 
![image](https://github.com/allo-protocol/aips/assets/22886639/b0189fdc-b1aa-4aff-b5d7-42e7e076a048)

By introducing an organization, builders can have a single place to handle their reputation and manage many projects or ideas. They can associate each project or idea with the organization and then apply them to the round. It enables a more organized and coherent system where each project or idea can be clearly identified, while also benefiting from the reputation associated with the organization.

@gravityblast  @meg-fsc @lmcorbalan wdyt?

**Reply by codenamejason** [reply]
I like the idea of an organization @KurtMerbeth then the reputation is much more manageable by the projects/ideas under it.

**Reply by thelostone-mc** [reply]
> I think that @lmcorbalan 's case is actually the "normal" case rather than an edge case for direct grants. It's very common for a direct grants program to have an open RFP (submit a great idea you want to build) and for projects to submit multiple applications/ideas for it.
> The way I see it, the meaning of a Project depends on the context, for QF Round it represents the combination of Builder+Idea and on the context of Direct Round it is just Builder and the Application is the Idea. Who knows what it would mean on a third context.

@lmcorbalan @meg-fsc  After reading that I  see the validity of having multiple applications. 
My only concern is that allowing a project to have different distinctly different applications makes things trickier in the following means: 

- Tracking the history of an idea and how it's progressed over rounds becomes trickier. This I think is something we would want to track as a project. Each project should have its own lifecycle/history to track IMO. 
- If anyone wanted to retrieve a project's application Id, assuming they know only know the project ID, they would have to go through all the applications and figuring out which is the right application might be trickier (we had this issue while @0xZakk was building out the sweepstake contract)
 
 If reputation is what we are worried about (which we should), I do like exploring having an organization contract which maybe adds the layer of unifying reputation. This might be worthwhile exploring as opposed to having a project to be able to submit multiple ideas. 
We'll take this into account when we review our design this week and make sure we account for this scenario.

**Reply by gravityblast** [reply]
I don't think we need another abstraction on top of the project. The organization with the reputation is already the Project itself, and instead of having multiple projects you have multiple applications. The application itself should have its purpose and description instead of cloning the current project snapshot. This will allow projects to update the application without having to update the project and then the application again. And projects can be updated without touching an application already submitted. 

> If anyone wanted to retrieve a project's application Id, assuming they know only know the project ID, they would have to go through all the applications and figuring out which is the right application might be trickier (we had this issue while @0xZakk was building out the sweepstake contract)

this would be the same if you have an org id, you would need to check all the projects and then all the applications. 

I like the idea of having Orgs more than Projects, but mostly as naming, and it makes more sense!
But in @lmcorbalan's case they would need an Org, and then a new project every time they apply to a round. 
Instead if we just have a project (or Org) and a specific metadata for each application would allow them to have a clear description of the project/org and a specific description for each round.

Moreover, I think having multiple applications or not should be an option of the contract, so round operators can decide if projects can apply multiple times or re-apply resetting their application status.

**Reply by meglister** [reply]
Agree with this approach @gravityblast

**Comment by gravityblast** [comment]
> In an ideal world, every project has just one status.

Having a clear distinction between Projects and Applications was a deliberate decisions we took when we decentralized gitcoin to avoid the same problems we had in cgrants.
In an ideal world actually Allo could apply to an Infra round with 2 applications: 1 to support zkrollups, 2 to support another chain. each application can receive votes and they have different descriptions, but the project and team is the same, because they need to bring their own reputation with both the applications.

**Comment by gravityblast** [comment]
> I propose introducing an extended set of application statuses to address the limitations of the current implementation. Here's a suggested set of application statuses using a 4-bit system:

I think we could go on for ever if we try to support all the possible statuses that a round operator can use.
For this reason I think the used bits for the status should be an arguments of the constructor. 
I propose to have a minimum of 1 bit (or 2), and the first 2 (or 4) statuses hardcoded in the system:

* `0` for `undefined` of `pending`
* `1` for `approved` 
* maybe also `2` and `3` like in the current system

At round creation the round operator can add more as long as 256 can be divided by the total number of statuses.
If people build on top of the protocol I'm sure they need pending/approved and maybe `rejected`, but they don't need other round specific statuses.
